guides/hack/21-XHP/16-extending.md

```diff
@@ -3,16 +3,29 @@ items that are not in the standard HTML specification.
 
 ## Basics
 
-All XHP class names start with a colon `:` and may include other `:` as well, as long as they are not adjacent. Note that this is an
-exception to the Hack rule where you cannot have `:` in class names.
-
-A custom XHP class needs to do two things:
-* extend `:x:element`.
-* implement the method `render` to return an XHP Object via `XHPRoot`.
+XHP class names must follow the same rules as any other Hack class names:
+Letters, numbers and `_` are allowed and the name mustn't start with a number.
+
+**Historical note:** Before XHP namespace support (in XHP-Lib v3), XHP class
+names could also contain `:` (now a namespace separator) and `-` (now disallowed
+completely). These were translated to `__` and `_` respectively at runtime (this
+is called "name mangling"). For example, `<ui:big-table>` would instantiate a
+global class named `xhp_ui__big_table`.
+
+A custom XHP class needs to do three things:
+* use the keywords `xhp class` instead of `class`
+* extend `x\element` (`\Facebook\XHP\Core\element`) or, rarely, another
+  [base class](/hack/XHP/interfaces)
+* implement the method `renderAsync` to return an XHP object (`x\node`) or the
+  respective method of the chosen base class
 
 @@ extending-examples/basic.inc.php @@
 @@ extending-examples/basic.php @@
 
+**Historical note:** Before XHP namespace support (in XHP-Lib v3), use
+`class :intro_plain_str` instead of `xhp class intro_plain_str` (no `xhp`
+keyword, but requires a `:` prefix in the class name).
```

Comment:
we probably want an fb span here saying that (for now) we're also using the old syntax, and no namespace support